### PR TITLE
Serialize and deserialize MapperSensorManager from karto::Mapper

### PR DESCRIPTION
**Pull request by ivonaj**
Write and read from a binary archive MapperSensorManager that contains the Pose Graph - LocalizedRangeScans

### CODE REVIEW

File: include/open_karto/Karto.h [hunk 1]
```diff
@@ -736,10 +738,10 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
 	template<class Archive>
 	void serialize(Archive &ar, const unsigned int version)
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: it looks like all the serialization methods are incorrectly tabbed over 1, can you get them back in line?

File: include/open_karto/Karto.h [hunk 2]
```diff
@@ -736,10 +738,10 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
 	{
-		ar & BOOST_SERIALIZATION_BASE_OBJECT_NVP(NonCopyable); 
+		ar & BOOST_SERIALIZATION_BASE_OBJECT_NVP(NonCopyable);
 		ar & BOOST_SERIALIZATION_NVP(m_Name);
-		ar & BOOST_SERIALIZATION_NVP(m_pParameterManager);
-	    //ar & boost::serialization::make_nvp("m_pParameterManager",*m_pParameterManager);
+//		ar & BOOST_SERIALIZATION_NVP(m_pParameterManager);
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: Do we not need the parameter manager?

File: include/open_karto/Karto.h [hunk 3]
```diff
@@ -736,10 +738,10 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
+//	    ar & boost::serialization::make_nvp("m_pParameterManager",*m_pParameterManager);
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: Feel  free to remove

File: include/open_karto/Karto.h [hunk 4]
```diff
@@ -3680,6 +3681,27 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
      */
     static SensorManager* GetInstance();
 
+  friend class boost::serialization::access;
+  template<class Archive>
+  void serialize(Archive &ar, const unsigned int version)
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: empty?

File: include/open_karto/Karto.h [hunk 5]
```diff
@@ -3790,6 +3812,10 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
      * Sensor map
      */
     SensorManagerMap m_Sensors;
+
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: spaces?

File: include/open_karto/Karto.h [hunk 6]
```diff
@@ -4309,6 +4335,12 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
     kt_int32u m_NumberOfRangeReadings;
 
     // static std::string LaserRangeFinderTypeNames[6];
+    friend class boost::serialization::access;
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: dempty?

File: include/open_karto/Karto.h [hunk 7]
```diff
@@ -5091,7 +5130,21 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
+  template<class Archive>
+  void serialize(Archive &ar, const unsigned int version)
+  {
+    ar & BOOST_SERIALIZATION_NVP(m_StateId);
+    ar & BOOST_SERIALIZATION_NVP(m_UniqueId);
+    ar & BOOST_SERIALIZATION_NVP(m_SensorName);
+    ar & BOOST_SERIALIZATION_NVP(m_Time);
+    ar & BOOST_SERIALIZATION_NVP(m_CustomData);
+    ar & BOOST_SERIALIZATION_BASE_OBJECT_NVP(Object);
+
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: extra line

File: include/open_karto/Karto.h [hunk 8]
```diff
@@ -5234,6 +5287,20 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
     kt_int32u m_NumberOfRangeReadings;
+
+  friend class boost::serialization::access;
+  template<class Archive>
+  void serialize(Archive &ar, const unsigned int version)
+  {
+    ar & BOOST_SERIALIZATION_NVP(m_NumberOfRangeReadings);
+    ar & BOOST_SERIALIZATION_BASE_OBJECT_NVP(SensorData);
+
+   if (Archive::is_loading::value)
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: please do brackets around if statements

File: include/open_karto/Mapper.h [hunk 9]
```diff
@@ -1356,12 +1356,12 @@ namespace karto
 	void serialize(Archive &ar, const unsigned int version)
 	{
 		//ar & boost::serialization::make_nvp("m_Values_1", m_Values[1]);
-		
 		ar & BOOST_SERIALIZATION_NVP(m_ScanManagers);
 		ar & BOOST_SERIALIZATION_NVP(m_RunningBufferMaximumSize);
 		ar & BOOST_SERIALIZATION_NVP(m_RunningBufferMaximumDistance);
 		ar & BOOST_SERIALIZATION_NVP(m_NextScanId);
 		ar & BOOST_SERIALIZATION_NVP(m_Scans);
+
```
Comment (SteveMacenski) [inline include/open_karto/Mapper.h]: extra line

File: src/Mapper.cpp [hunk 10]
```diff
@@ -2207,23 +2208,44 @@ namespace karto
       m_Initialized = true;
 	}
   }
-
-  void Mapper::SaveToFile(const std::string& filename) 
+  void Mapper::SaveToFile(const std::string& filename)
```
Comment (SteveMacenski) [inline src/Mapper.cpp]: removed space?

File: src/Mapper.cpp [hunk 11]
```diff
@@ -2207,23 +2208,44 @@ namespace karto
...
   {
 	  printf("Save To File\n");
 	  std::ofstream ofs(filename.c_str());
-	  boost::archive::xml_oarchive oa(ofs);
-	  //save class state to archive
-	  oa << BOOST_SERIALIZATION_NVP(m_pMapperSensorManager);
-  }	
+	  boost::archive::binary_oarchive oa(ofs, boost::archive::no_codecvt);//using binary archive because it has support for NanN values	  //save class state to archive
+//    boost::archive::binary_oarchive oa(ofs);
```
Comment (SteveMacenski) [inline src/Mapper.cpp]: are these needed?

File: src/Mapper.cpp [hunk 12]
```diff
@@ -2207,23 +2208,44 @@ namespace karto
...
   {
     printf("Load From File\n");
 	  std::ifstream ifs(filename.c_str());
-	  boost::archive::xml_iarchive ia(ifs);
+//    boost::archive::binary_iarchive ia(ifs);
+	  boost::archive::binary_iarchive ia(ifs, boost::archive::no_codecvt);
 	  //read class state from archive
-	  ia >> BOOST_SERIALIZATION_NVP(m_pMapperSensorManager);
+
+//    ia.template register_type<MapperSensorManager>();
```
Comment (SteveMacenski) [inline src/Mapper.cpp]: are these needed?

File: include/open_karto/Karto.h [hunk 13]
```diff
@@ -5497,6 +5585,21 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
       }
     }
 
+    inline void SetPointReadings(PointVectorDouble& points, kt_bool setFiltered = false)
+    {
+        if (setFiltered == true)
+
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: space

File: include/open_karto/Karto.h [hunk 14]
```diff
@@ -5497,6 +5585,21 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
       }
     }
 
+    inline void SetPointReadings(PointVectorDouble& points, kt_bool setFiltered = false)
+    {
+        if (setFiltered == true)
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: you can just do `if(setFiltered)`

File: src/Mapper.cpp [hunk 15]
```diff
@@ -2207,23 +2208,44 @@ namespace karto
...
+    LocalizedRangeScan *d;
+    Pose2 one_scan;
+    kt_int32u name;
+
+    CustomDataVector cus;
+    std::vector<kt_double> rrvec;
+    kt_double reading;
+    ia >> BOOST_SERIALIZATION_NVP(mapper);
+    m_pMapperSensorManager = mapper;
+
```
Comment (SteveMacenski) [inline src/Mapper.cpp]: space

File: src/Karto.cpp [hunk 16]
```diff
@@ -26,11 +26,20 @@
...
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::Object);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::Sensor);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::SensorData);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::Name);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::LaserRangeScan);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::LocalizedRangeScan);
+BOOST_CLASS_EXPORT_IMPLEMENT(karto::SensorManager);
+//BOOST_CLASS_EXPORT_IMPLEMENT(karto::LaserRangeFinder);
+
+
```
Comment (SteveMacenski) [inline src/Karto.cpp]: spaces

File: include/open_karto/Karto.h [hunk 17]
```diff
@@ -6817,10 +6919,30 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
+
+BOOST_CLASS_EXPORT_KEY(karto::SensorData);
+BOOST_CLASS_EXPORT_KEY(karto::LocalizedRangeScan);
+BOOST_CLASS_EXPORT_KEY(karto::LaserRangeScan);
+
+
+
+
+
+
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: spaces

File: include/open_karto/Karto.h [hunk 18]
```diff
@@ -5338,6 +5405,11 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
     mutable boost::shared_mutex m_Lock;
 
   public:
+      inline void callUpdate()
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: who's the client of this?

Comment (ivonaj) [reply include/open_karto/Karto.h]: isn't used, I'll delete it

File: include/open_karto/Karto.h [hunk 19]
```diff
@@ -5425,10 +5463,16 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
...
+    inline void SetIsDirty(kt_bool& rIsDirty)
+    {
+        m_IsDirty = rIsDirty;
+    }
+
     /**
      * Computes the robot pose given the corrected scan pose
      * @param rScanPose pose of the sensor
      */
+
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: remove space

File: include/open_karto/Karto.h [hunk 20]
```diff
@@ -5497,6 +5546,20 @@ BOOST_SERIALIZATION_ASSUME_ABSTRACT(NonCopyable)
       }
     }
 
+    inline void SetPointReadings(PointVectorDouble& points, kt_bool setFiltered = false)
+    {
+      if (setFiltered)
+      {
+        m_PointReadings.clear();
```
Comment (SteveMacenski) [inline include/open_karto/Karto.h]: kill line

File: src/Mapper.cpp (hunk 10 above)
Comment (SteveMacenski) [reply src/Mapper.cpp]: tab issues

**Pull request merged**